### DISCUSSION TITLE
Can I customize the behaviour of the DELETE button to soft delete instead?

### DISCUSSION THREAD

**Discussion by glaucocustodio** [avo-hq/avo#2550] [Help] [answered]
Hey guys, is that possible to override the default behaviour of the delete button?

I'm using the https://github.com/jhawthorn/discard gem and would like to call `discard` on the record instead of `destroy`.

If overriding is not possible, can I hide the delete button at least so I can only keep a discard action visible?

**Comment by adrianthedev** [comment, marked as answer]
There are a few ways of doing that.

1. Use [`show_controls`](https://docs.avohq.io/3.0/customizable-controls.html#show_page) to hide the destroy button and add the soft-delete one on the show page
2. Use the [`authorization`](https://docs.avohq.io/3.0/authorization.html#destroy) feature to remove the ability to destroy a record for a user

**Reply by glaucocustodio** [reply]
Thanks for the response @adrianthedev, I've chosen to use the authorization feature:
```ruby
class ProjectPolicy < ApplicationPolicy
  def index? = true

  def show? = true

  def create? = true

  def new? = true

  def update? = true

  def edit? = true

  def destroy? = false

  class Scope < Scope
    def resolve
      scope.all
    end
  end
end
```
In development env it works fine (I don't see the bin icon), but in production I still see the bin icon, not sure why 🤔.
When I point my local application to production (`RAILS_ENV=production bin/dev`) I also see the bin icon.
I am trying to find out why by having a look at avo source code but haven't found the reason yet. Any clues?

**Reply by adrianthedev** [reply]
What license type are you using?
Authorization is a [Pro feature](https://avohq.io/pricing).

If you are on Pro or Advanced, then ensure you have the [license token added](https://docs.avohq.io/3.0/licensing.html) to the production environment too.

**Reply by glaucocustodio** [reply]
Sorry, we were using the old Avo 2 licence key, it has been sorted now, thank you.

**Comment by adrianthedev** [comment]
3. We'll take a PR to add support for a soft-delete button 👈 the best solution IMHO

**Comment by kira4ka** [comment]
@adrianthedev does this [discussion](https://github.com/avo-hq/avo/pull/1818) may be implemented here? 
Because it seems that there is already a custom behaviour but you are saying to use different approach.

**Reply by Paul-Bob** [reply]
@kira4ka overriding the `destroy_record_action` method on the desired controller should work perfectly for this scenario. Nice spot! Thanks for sharing it.

**Comment by adrianthedev** [comment]
Let's start a conversation here about how to implement the soft-delete (calling it SD from now on)  functionality in Avo.

In the begining it looks like a simple problem that a developer only needs to add the soft-delete action to the UI, but it goes a bit deeper: 

- how to scope the query to view the right records
- how to show the soft-deleted records
- how to show all the record
- how to configure to whom to show them (authorization)
- how to configure who can SD records (authorization)


My thoughts:

 - the SD (soft-delete) button should be hidden by default
 - the SD action should not be enabled by default
 - as there isn't a native way to do this, it should be easy to use one of the gems that make this possible
 - support custom implementations (this is already possible using custom controls and query modification)


### Possible gems

 - https://github.com/jhawthorn/discard
 - https://github.com/grosser/soft_deletion
 - https://github.com/rubysherpas/paranoia
 - https://github.com/ActsAsParanoid/acts_as_paranoid

# My gem recommendations

It seems like `discard` is the latest and has the most learnings from all the others.

I wouldn't use `soft_deletion` because of the API. No timestamp on SD column.

`paranoia` is not accepting new features, so wouldn't support it as a first-party gem.

`paranoid` doesn't seem to work with Rails 7 and onward.

https://www.ruby-toolbox.com/categories/Active_Record_Soft_Delete


I am looking for more things I might have overlooked. Let's keep this conversation going.

**Reply by kira4ka** [reply]
Discard gem took over from of act_as_paranoid and paranoia. So, I think that for the moment this will be the most requested integration.

`how to scope the query to view the right records` -> The scope is set on model. And AVO manages it well, as it doesn't show discarded records when  using `default_scope -> { kept } ` and shows them differently otherwise.
`how to show the soft-deleted records`` how to show all the record` ->  Avo also allows the use of scoped views. So, If I want to see all records or not I have this possibility with scope function. Discard allows the use of different scopes such as kept, as discarded end so on. 
`how to configure to whom to show them (authorization)` `how to configure who can SD records (authorization)` ->  Simply working on Pundit policies should be enough for authorization. 

`the SD (soft-delete) button should be hidden by default ` -> Mmmh. why? For the end user this part should be transparent, as from the moment that the user wants to delete a record, they don't care about what's happening with this record on DB level. And in my opinion, if my model includes `Discard::Model` it means that my delete button should perform a SD. 
It might be disabled by default, but having a quick option to enable it (without the need to modify actions or overriding methods) would be beneficial.
